src/main/java/org/opensearch/neuralsearch/processor/factory/RerankProcessorFactory.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 Aryn
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.opensearch.neuralsearch.processor.factory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.AllArgsConstructor;
+
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+import org.opensearch.neuralsearch.processor.rerank.ContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.DocumentContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.QueryContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.RerankType;
+import org.opensearch.neuralsearch.processor.rerank.TextSimilarityRerankProcessor;
+import org.opensearch.search.pipeline.Processor;
+import org.opensearch.search.pipeline.SearchResponseProcessor;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Factory for rerank processors. Must:
+ * - Instantiate the right kind of rerank processor
+ * - Instantiate the appropriate context source fetchers
+ */
+@AllArgsConstructor
+public class RerankProcessorFactory implements Processor.Factory<SearchResponseProcessor> {
+
+    public static final String RERANK_PROCESSOR_TYPE = "rerank";
+    public static final String CONTEXT_CONFIG_FIELD = "context";
+
+    private final MLCommonsClientAccessor clientAccessor;
+
+    @Override
+    public SearchResponseProcessor create(
+        final Map<String, Processor.Factory<SearchResponseProcessor>> processorFactories,
+        final String tag,
+        final String description,
+        final boolean ignoreFailure,
+        final Map<String, Object> config,
+        final Processor.PipelineContext pipelineContext
+    ) {
+        RerankType type = findRerankType(config);
+        boolean includeQueryContextFetcher = ContextFetcherFactory.shouldIncludeQueryContextFetcher(type);
+        List<ContextSourceFetcher> contextFetchers = ContextFetcherFactory.createFetchers(config, includeQueryContextFetcher);
+        switch (type) {
+            case TEXT_SIMILARITY:
+                @SuppressWarnings("unchecked")
+                Map<String, String> rerankerConfig = (Map<String, String>) config.remove(type.getLabel());
```

Comment:
Better ways to parse the config for the processors.

https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java#L52

Reply:
ack. will use these everywhere if possible